Add VP12 SCS Health Monitor blog post (#992)
SovereignCloudStack/website: _i18n/en/_posts/blog/2024-08-20-vp12-scs-health-monitor-tech-preview.md

---
layout: post
title:  "SCS Health Monitor released: What's inside and how to use it"
category: tech
author:
- "Piotr Bigos"
- "Katharina Trentau"
- "Tomáš Smädo"
- "Dominik Pataky"
avatar:
- "piobig2871.jpg"
- "default-avatar.jpg"
- "bitkeks.png"
about:
- "piobig2871"
- "bitkeks"
---

Coming closer to SCS release R7, more and more projects are finalized and ready to be tested by a broader audience. This also includes the newly developed SCS Health Monitor (SCSHM), which we present to you in this blog post!

## Motivation

To start, what exactly do we mean by "health monitoring"? Aren't metrics, monitoring dashboards and alerts good indicators of healthy platforms? – While this might be enough in general, SCS also takes another aspect of "health" into account: actively testing behaviour of the platform.

> This means automated systems execute predefined actions, using the same credentials as users would, against SCS-based systems. Running these tests covers many paths and workflows end-customers or operators would also take in their daily doing, so that tests are very much aligned with real-world use cases. Errors in the platform can be caught early on. Some might know this pattern as "Behavior Driven Development" (BDD) and BDD testing.

Reviewer: ```suggestion

This means automated systems execute predefined actions, using the same credentials as users would, against SCS-based systems. Running these scenario tests covers many paths and workflows end-customers or operators would also take in their daily doing, so that tests are very much aligned with real-world use cases. Errors in the platform can be caught early on. Some might know this pattern as "Behavior Driven Development" (BDD), BDD testing or Behavior Driven Monitoring.
```
Calling this a scenario test and reference the Behavior Driven Monitoring term that we have used before.

Reviewer 2: reads good @garloff

## Previous work

> Even though we are talking about a "tech preview", the concept for BDD testing in SCS is not new. For some time now, the [OpenStack Health Monitor](https://github.com/SovereignCloudStack/openstack-health-monitor/) (OSHM) runs and tests our partner's SCS installations and builds reports on Grafana dashboards. With the [initial commit](https://github.com/SovereignCloudStack/openstack-health-monitor/commit/1de625b9b5534235a92e925226adea5a3c2d7507) made in November 2019 by SCS CTO [Kurt Garloff](https://github.com/garloff), the OSHM has a long history and went through multiple iterations of patches, mainly submitted by Kurt, but also by others from the community. The broad deployment of the OSHM in CSP sites offers insights that everyone benefits from. – And they are publicly available from the dashboards, which are linked on the [overview of "Compliant cloud environments"](https://docs.scs.community/standards/certification/overview) page, column *HealthMon*.

Reviewer: ```suggestion
Even though we are talking about a "tech preview", the concept for BDD testing in SCS is not new. For some time now, the [OpenStack Health Monitor](https://github.com/SovereignCloudStack/openstack-health-monitor/) (OSHM) runs and tests our partner's SCS installations and builds reports on Grafana dashboards. The OSHM has a long history and was once created for the OpenTelekomCloud, modelling a CloudFoundry deployment that was reported to no work reliably. The [initial commit](https://github.com/SovereignCloudStack/openstack-health-monitor/commit/1de625b9b5534235a92e925226adea5a3c2d7507) into SCS was made in November 2019 by SCS CTO [Kurt Garloff](https://github.com/garloff). The OSHM went through multiple iterations of patches, mainly submitted by Kurt, but also by others from the community. The broad deployment of the OSHM in CSP sites offers insights that everyone benefits from. – And they are publicly available from the dashboards, which are linked on the [overview of "Compliant cloud environments"](https://docs.scs.community/standards/certification/overview) page, column *HealthMon*.
```
The history goes back to 2016 or 2017, actually.

## Concept

The new approach should lay a ground for testing not only OpenStack (a limiting disadvantage of both the OSHM and StackMon), but Kubernetes and applications on top as well. Originally, the team for the SCSHM work package (tender VP12) brought in their own technical proposal for a new implementation of a health monitoring stack. It was decided that this design was then used as a blueprint for all SCS components and layers, including OpenStack.

At the time of writing, the implementation of the SCS Health Monitor makes use of:
* the [Gherkin DSL](https://cucumber.io/docs/gherkin/reference/) in which test cases are written in a non-technical human-readable language,
* the [Behave](https://behave.readthedocs.io/en/latest/)/[BehaveX](https://github.com/hrcorval/behavex) libraries, which are written in Python and translate Gherkin into code,
* API clients for [OpenStack](https://docs.openstack.org/openstacksdk/latest/user/index.html), [Kubernetes](https://kubernetes.readthedocs.io/en/latest/) and any other application, and
* a Prometheus-based setup for exporting metrics.

> Given that the implementation of API clients is done in Python, the framework is extendable to other tools as well (IAM, security, secrets, CI/CD, system monitoring). Many APIs in the SCS ecosystem work via HTTPS by default. Where there's no Python SDK or client library, we can fall back to plain HTTP requests.

Reviewer: ```suggestion
Given that the implementation of API clients is done in Python, the framework is extendable to other tools as well (IAM, security, secrets, CI/CD, system monitoring). Many APIs in the SCS ecosystem work via HTTPS by default. Where there's no Python SDK or client library, we can fall back to plain HTTPS requests.
```
Avoid raised eyebrows for "http" really without "s"?
